adlfs/transfer.py

```diff
@@ -0,0 +1,312 @@
+            temporary chunk.
+        tmp_unique: bool [True]
+            If True, then a unique ID will be generated to create a subdirectory
+            containing the temporary chunks. Otherwise, all temporary chunks
+            will be placed in `tmp_path`.
+        persist_path: str [None]
+            Path used for persisting a client's state. If None, then `save()`
+            and `load()` will be empty operations.
+        delimiter: byte(s) or None
+            If set, will transfer blocks using delimiters, as well as split
+            files for transferring on that delimiter.
+        """
+        self._adlfs = adlfs
+        self._name = name
+        self._transfer = transfer
+        self._merge = merge
+        self._nthreads = nthreads or multiprocessing.cpu_count()
+        self._chunksize = chunksize
+        self._chunkretries = 5
+        self._tmp_path = tmp_path
+        self._tmp_prefix = tmp_prefix
+        self._tmp_unique = tmp_unique
+        self._persist_path = persist_path
+        self._pool = ThreadPoolExecutor(self._nthreads)
+        self._shutdown_event = threading.Event()
+        self._files = {}
+        self._fstates = DisjointState(
+            'pending', 'transferring', 'merging', 'finished', 'cancelled',
+            'errored')
+
+    def submit(self, src, dst, nbytes):
+        """
+        All submitted files start in the `pending` state until `run()` is
+        called.
+        """
+        self._fstates[(src, dst)] = 'pending'
+        self._files[(src, dst)] = dict(
+            nbytes=nbytes,
+            start=None,
+            stop=None,
+            chunks={},
+            cstates=DisjointState('running', 'finished', 'cancelled', 'errored'),
+            merge=None)
+
+    def _submit(self, fn, *args, **kwargs):
+        kwargs['shutdown_event'] = self._shutdown_event
+        return self._pool.submit(fn, *args, **kwargs)
+
+    def _scatter(self, src, dst, transfer):
+        """ Split a given file into chunks """
+        dic = self._files[(src, dst)]
+        self._fstates[(src, dst)] = 'transferring'
+        offsets = list(range(0, dic['nbytes'], self._chunksize))
+        for offset in offsets:
+            if self._tmp_path and len(offsets) > 1:
+                subdir = uuid.uuid1().hex[:10] if self._tmp_unique else ''
+                prefix = self._tmp_prefix or ''
+                name = os.path.join(
+                    self._tmp_path,
+                    subdir,
+                    prefix + str(offset))
+            else:
+                name = dst
+            logger.debug("Submitted %s, byte offset %d", name, offset)
+            dic['cstates'][name] = 'running'
+            dic['chunks'][name] = dict(
+                future=self._submit(transfer, self._adlfs, src, name, offset,
+                                    self._chunksize),
+                retries=self._chunkretries,
+                offset=offset)
+
+    @property
+    def progress(self):
```

Comment:
Agree. Not sure if we need to track total time in here (since that is easy for someone to measure externally) but the important things to track are:
Total File Size (in bytes)
Bytes Uploaded
Total chunks
Chunks remaining.

---
In reply to: [78791135](https://github.com/Azure/azure-data-lake-store-python/pull/64#discussion_r78791135) [](ancestors = 78791135)